Add blog Debezium connector blog (#115)
strimzi/strimzi.github.io: _posts/2019-11-01-deploying-debezium-with-kafkaconnector-resource.md

---
layout: post
title:  "Deploying Debezium using the new KafkaConnector resource"
date: 2019-11-01
author: tom_bentley
---

For too long our Kafka Connect story hasn't quite as "Kubernetes-native" as it could have been.
We had a `KafkaConnect` resource to configure a Kafka Connect _cluster_ but you still had to use the Kafka Connect REST API to actually create a _connector_ within it.
While this wasn't especially difficult using something like `curl`, it stood out because everything else could be done using `kubectl` and it meant that connectors didn't fit into our Kubernetes-native vision.
With the help of a contribution from the community, Strimzi now supports a `KafkaConnector` custom resource and the rest of this blog post is going to explain how to use it using [Debezium](https://debezium.io) as an example. 

> And as if that wasn't enough, there's some awesome ASCII art.

Reviewer: I don't quite get this one.

Author: Yeah, it's a weak attempt to add a little humour. Did you appreciate the joke in the end (maybe I need to tweak this first sentence to make it work better)? Or did you think it was too lame to be worth it?

Reviewer: Yep, got it in the end and liked it. It won't nominate you for the comedy Oscars, but it's not lame either. It was just a bit contextless here, perhaps make some forward-looking link to qualify it a bit. Perhaps "How so? Read on to find out" or similar would do the trick.

<!--more-->

# Debezi-what?

> If you've not heard of Debezium before it is an open source project for applying [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture) (CDC) pattern to your applications using Kafka.

Reviewer: ```suggestion
If you've not heard of Debezium before it is an open source project for applying the [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture) (CDC) pattern to your applications using Kafka.
```

But what is CDC? You probable have several databases in your organization; silos full of business-critical data.
While you can query those databases any time you want, the essence of your business revolves around how that data gets modified.
Those modifications trigger, and are triggered by, real world business events (e.g. a phone call from a customer, or a successful payment or whatever)
In order to reflect this, modern application architectures are frequently event-based. 
And so asking for the _current state_ of some tables is not enough; what these architectures need is a stream of _events representing modifications_ to those tables.
That's what CDC is: Capturing the changes to the state data as event data.

Concretely, Debezium works with a number of common DBMSs (MySQL, MongoDB, PostgreSQL, Oracle, SQL Server and Cassandra) and runs as a source connector within a Kafka Connect cluster.
How Debezium works on the database side depends which database it's using.

> For example for MySQL it reads the commit log in order to know what transaction are happening, but for MongoDB it hooks into the native replication mechanism.

Reviewer: ```suggestion
For example for MySQL it reads the commit log in order to know what transactions are happening, but for MongoDB it hooks into the native replication mechanism.
```

> In any case, the changes get represented as JSON events (using a common schema) which are sent to a Kafka.

Reviewer: JSON is just one particular event serialization, it could also be Avro or yet something else.

s/a Kafka/Kafka/

It should be apparent, then, that Debezium provides a route for getting events out of database applications (which otherwise might not expose any kind of event-based API) and make them available to Kafka applications.

So that's what Debezium _is_ and what it _does_. Its role in this blog post is just to be an example connector to use with the `KafkaConnector` which is new in Strimzi 0.15.0.

Let's get cracking with the walkthrough...

# Fire up MySQL

Let's follow the steps from the [Debezium tutorial](https://debezium.io/documentation/reference/0.10/tutorial.html) for getting a demo MySQL server up and running.

First we fire up the database server in a Docker container:

```shell
docker run -it --rm --name mysql -p 3306:3306 \
  -e MYSQL_ROOT_PASSWORD=debezium -e MYSQL_USER=mysqluser \
  -e MYSQL_PASSWORD=mysqlpw debezium/example-mysql:0.10
```

> Once we see the following we know the server is up and running:

Reviewer: "up and running" is repeated with the occurence just above.

```shell
...
017-09-21T07:18:50.824629Z 0 [Note] mysqld: ready for connections.
Version: '5.7.19-log'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server (GPL)
```

Then we can run the command line client:

```shell
docker run -it --rm --name mysqlterm --link mysql --rm mysql:5.7 sh \
  -c 'exec mysql -h"$MYSQL_PORT_3306_TCP_ADDR" \
  -P"$MYSQL_PORT_3306_TCP_PORT" \
  -uroot -p"$MYSQL_ENV_MYSQL_ROOT_PASSWORD"'
```

At the `mysql>` prompt we can switch to the "inventory" database and show the tables in it:

```shell
mysql> use inventory;
mysql> show tables;
```

The output will look like this:

```shell
+---------------------+
| Tables_in_inventory |
+---------------------+
| customers           |
| orders              |
| products            |
| products_on_hand    |
+---------------------+
4 rows in set (0.00 sec)
```

> > Don't worry, this isn't the awesome ascii art.

Reviewer: ```suggestion
> Don't worry, this isn't the awesome ASCII art.
```

> If you want, you can have a poke around this demo database, but when you're done leave this mySQL client running in its own terminal window, so you can come back to it later.

Reviewer: ```suggestion
If you want, you can have a poke around this demo database, but when you're done leave this MySQL client running in its own terminal window, so you can come back to it later.
```

```

And spin up a Kafka cluster, waiting until it's ready:

```shell
kubectl -n kafka \
    apply -f https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/0.15.0/examples/kafka/kafka-persistent-single.yaml \
  && kubectl wait kafka/my-cluster --for=condition=Ready --timeout=300s -n kafka
```

What we've got so far looks like this

<pre style="line-height: 1.2 !important;"><code>
┌────────────────────────────────┐         ┌────────────────────────────────┐
│ minikube, namespace: kafka     │         │ docker                         │
│                                │         │                                │
│ Kafka                          │         │ MySQL                          │
│  name: my-cluster              │         │                                │
│                                │         └────────────────────────────────┘
│ KafkaConnect                   │
│  name: my-connect              │
│                                │
└────────────────────────────────┘
</code></pre>

> Don't worry, the ASCII art gets a lot better than this!

What's missing from this picture is Kafka Connect in the minikube box.

# Kafka Connect image

> The next step is to [create a Strimzi Kafka Connect image](https://strimzi.io/docs/master/#creating-new-image-from-base-str) with the Debezium connector jar file. 

Reviewer: jar=archive? In any case, it's multiple JARs.

First download and extract the Debezium MySQL connector archive
```shell
curl https://repo1.maven.org/maven2/io/debezium/debezium-connector-mysql/0.10.0.Final/debezium-connector-mysql-0.10.0.Final-plugin.tar.gz \
| tar xvz
```

Prepare a `Dockerfile` which adds those connector files to the Strimzi Kafka Connect image

```shell
cat <<EOF >Dockerfile
FROM strimzi/kafka:latest-kafka-2.3.0
USER root:root
EXEC mkdir -p /opt/kafka/plugins/debezium
COPY ./debezium-connector-mysql/ /opt/kafka/plugins/debezium/
USER 1001
EOF
```

> Then build the image from that `Dockerfile` and push it to dockerhub.

Reviewer: Is this remote push strictly required, or can it be done locally, too? If needed indeed, a sentence is warranted that an account is needed, but it adds quite a bit of friction.

Author: It is possible to do this (e.g. [`eval $(minikube docker-env)`](https://stackoverflow.com/questions/42564058/how-to-use-local-docker-images-with-minikube#42564211)), but that method doesn't work because we have no control over the imagePullPolicy. Other ways of achieving the same thing are more complex, so in the end I felt that going via dockerhub, while not ideal, was simple and guaranteed to work. Maybe @scholzj knows of a better way of doing it?

```shell
# You can use your own dockerhub organization
export DOCKER_ORG=tjbentley
docker build . -t ${DOCKER_ORG}/connect-debezium
docker push ${DOCKER_ORG}/connect-debezium
```

# Create the Connect cluster

Now we can create a `KafkaConnect` cluster in Kubernetes:

```shell
cat <<EOF | kubectl apply -f -
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaConnect
metadata:
  name: my-connect-cluster
  annotations:
  # use-connector-resources configures this KafkaConnect
  # to use KafkaConnector resources to avoid
  # needing to call the Connect REST API directly
    strimzi.io/use-connector-resources: "true"
spec:
  image: ${DOCKER_ORG}/connect-debezium
  replicas: 1
  bootstrapServers: my-cluster-kafka-bootstrap:9093
  tls:
    trustedCertificates:
      - secretName: my-cluster-cluster-ca-cert
        certificate: ca.crt
EOF
```

It's worth pointing out a couple of things about the above resource:

> * The `strimzi.io/use-connector-resources: "true"` annotation tells the cluster operator that this `KafkaConnect` is managed via `KafkaConnector` resources.

Reviewer 2: I don't get the point of this sentence. Anyway, the CO is going to deploy a Kafka Connect cluster via the `KafkaConnect` resource; then you can create connectors via `KafkaConnector`. What do you mean by "this `KafkaConnect` is managed via `KafkaConnector` resources" ... I missed the point of "is managed via".

Author: Would "will be managed using `KafkaConnector` resources." be clearer @ppatierno ?

Reviewer 2: What sounds weird to me (which could actually be related to not being a native English) is that you want to say that `KafkaConnect` instance is managed via/using `KafkaConnector` resources. I think that from your point of view,  "managing" a Kafka Connect cluster means add/removing/configuring Kafka connectors so the "management" isn't via HTTP API but via connector resources. I see your point but it really sounds weird to me (again I don't know how it sounds in English). I would just highlight that connectors are created/deleted/configured in Kafka Connect via Kafka Connectors resources.

Author: So how about this: 
> The `strimzi.io/use-connector-resources: "true"` annotation tells the cluster operator that `KafkaConnector` resources will be used to configure connectors within this Kafka Connect cluster.

Reviewer 2: That sounds great to me!

* The `spec.image` is the image we just created with `docker`.

# Create the connector

The last piece is to create the `KafkaConnector` resource configured to connect to our "inventory" database in MySQL.

Here's what the `KafkaConnector` resource looks like:

> ```shell

Reviewer: Nice! Perhaps also show the password secret stuff we were discussing. That'd add some more meat on the actual new CRD feature this post is about and it nicely matches the DBZ use case.

Author: I'm so stupid :disappointed:. I guess this is the point where I _finally_ figure out _exactly_ why we shouldn't use the map-of-key-to-value approach which we've been flip-flopping over the whole time: Although we can mount a Secret into the container in the `KafkaConnect`, we still need to actually configure that parameter in the `KafkaConnector`. If it's mounted as a file we need a way to say "take the value from this file" and if it was set as an env var we still need a way to say "take the value from this env var". So I guess I need to change the `KafkaConnectorSpec` to support this before we merge the PR, right @scholzj @ppatierno ? So the only question is whether it should be a list of 

```
key:
value: #or valueFrom
```

Or a map with parameter name as key and value being an object with `value` and `valueFrom` properties. Since the list approach is consistent with Kubernetes, I propose to do it that way.

Reviewer 2: When we had this discussion I was more for having the consistent Kubernetes approach anyway, even if tbh I didn't recall as you the problem about mounting value from :-) Anyway I am for the list approach.

spec:
  class: io.debezium.connector.mysql.MySqlConnector
  tasksMax: 1
  config:
    database.hostname: 192.168.99.1
    database.port: "3306"
    database.user: "debezium"
    database.password: "dbz" 
    database.server.id: "184054"
    database.server.name: "dbserver1"
    database.whitelist: "inventory"
    database.history.kafka.bootstrap.servers: "my-cluster-kafka-bootstrap:9092"
    database.history.kafka.topic: "schema-changes.inventory"
    include.schema.changes: "true" 
EOF
```

In `metadata.labels`, `strimzi.io/cluster` names the `KafkaConnect` cluster which this connector will be created in.

The `spec.class` names the Debezium MySQL connector and `spec.tasksMax` can be 1 because that's all this connector ever uses.

The `spec.config` object contains the rest of the connector configuration.
The [Debezium documentation](https://debezium.io/documentation/reference/0.10/connectors/mysql.html#connector-properties) explains the available properties, but it's worth calling out some specifically:
* I'm using `database.hostname: 192.168.99.1` as IP address for connecting to MySQL because I'm using `minikube` with the virtualbox VM driver

>   If you're using a different VM drvier with `minikube` you might need a different IP address.

Reviewer: ```suggestion
  If you're using a different VM driver with `minikube` you might need a different IP address.
```

```
...
status:
  conditions:
  - lastTransitionTime: "2019-11-18T15:01:53.135Z"
    status: "True"
    type: Ready
  observedGeneration: 1
```

This tells us that the connector is running within the `KafkaConnect` cluster we created in the last step.

To summarise, we've now the complete picture with the connector talking to MySQL:

<pre style="line-height: 1.2 !important;"><code>
┌────────────────────────────────┐         ┌────────────────────────────────┐
│ minikube, namespace: kafka     │         │ docker                         │
│                                │         │                                │
│ Kafka                          │     ┏━━━┿━▶ MySQL                        │
│  name: my-cluster              │     ┃   │                                │
│                                │     ┃   └────────────────────────────────┘
│ KafkaConnect                   │     ┃
│  name: my-connect              │     ┃
│                                │     ┃
│ KafkaConnector                 │     ┃
│  name: inventory-connector ◀━━━┿━━━━━┛
│                                │
└────────────────────────────────┘
</code></pre>

> > OK, I admit it: I was lying about the ASCII art being awesome.

Reviewer: :)

I kinda like it.

# Showtime!

If you list the topics, for example using `kubectl exec my-cluster-kafka-0 -c kafka -i -t -- bin/kafka-topics.sh --zookeeper '' --list` you should see:

```
__consumer_offsets
connect-cluster-configs
connect-cluster-offsets
connect-cluster-status
dbserver1
dbserver1.inventory.addresses
dbserver1.inventory.customers

> dbserver1.inventory.geom

Reviewer: That's not in the (4) tables you were showing before?

Author: I didn't add that table, but it's in Debezium's demo MySQL dockers images. Do you know what it's there for?

Reviewer: Yes, it is to show off geospatial column types being captured. I was only wondering why it didn't show up in the four tables you listed somewhere before.

Author: Because I copy-and-pasted that bit from the Debezium tutorial. Now fixed.

dbserver1.inventory.orders
dbserver1.inventory.products
dbserver1.inventory.products_on_hand
schema-changes.inventory
```

The `connect-cluster-*` topics are the usual internal Kafka Connect topics.
Debezium has created a topic for the server itself (`dbserver1`), and one for each table within the `inventory` database.

Let's start consuming from one of those change topics:

```

> kubectl exec my-cluster-kafka-0 -c kafka -i -t -- \

Reviewer: For Debezium examples we moved to kafkacat lately, as it's much faster to start and also allows for nicer formatting/filtering via jq. E.g. see here if you're interested: https://github.com/debezium/debezium-examples/tree/master/auditlog#modifying-some-data-and-observing-the-audit-log

We provide a container image, which contains these tools, e.g. also mycli and pgcli as nicer CLI clients for the DBs. Just some food for thought.

Author: Strimzi's existing images don't include kafkacat. I not keen to distract from the main thrust of the content by having to deploy some other pod just to show the records in the topics in a marginally better way. Maybe exposing the cluster and accessing from outside would be less distracting.

But I take your point (and it applies not just to this blog, but to other blogs and to demos) that it's a bit awkward showing messages in topics with Strimzi at the moment. The whole `exec`ing in the broker pod is ugly and it misleads people into thinking that this is the correct way to do it.

Reviewer: Sure. I guess it's a bit more difficult to do indeed with K8s where a pod needs to be deployed. In our READMEs we simply use `docker run` with out tooling image and then kafkacat etc., so it doesn't add any overhead in terms of perceived complexitiy.

```

> which is rather a lot of JSON, but if we reformat it (e.g. using copying, paste and `jq`) we see:

Reviewer: I'd probably omit the schema, as it adds lots of nice. That's the nice thing of the kafkacat + jq combo, it lets you do this interactively.

            "optional": true,
            "field": "query"
          }
        ],
        "optional": false,
        "name": "io.debezium.connector.mysql.Source",
        "field": "source"
      },
      {
        "type": "string",
        "optional": false,
        "field": "op"
      },
      {
        "type": "int64",
        "optional": true,
        "field": "ts_ms"
      }
    ],
    "optional": false,
    "name": "dbserver1.inventory.customers.Envelope"
  },
  "payload": {
    "before": {
      "id": 1004,
      "first_name": "Anne",
      "last_name": "Kretchmar",
      "email": "annek@noanswer.org"
    },
    "after": {
      "id": 1004,
      "first_name": "Anne Mary",
      "last_name": "Kretchmar",
      "email": "annek@noanswer.org"
    },
    "source": {
      "version": "0.10.0.Final",
      "connector": "mysql",
      "name": "dbserver1",
      "ts_ms": 1574090237000,
      "snapshot": "false",
      "db": "inventory",
      "table": "customers",
      "server_id": 223344,
      "gtid": null,
      "file": "mysql-bin.000003",
      "pos": 4311,
      "row": 0,
      "thread": 3,
      "query": null
    },
    "op": "u",
    "ts_ms": 1574090237089
  }
}
```

> The `schema` object is just describing the schema of the rest of the event.

Reviewer: ```suggestion
The `schema` object is describing the schema of the actual event payload.
```

> What's more interesting for this post is the `payload`. Working backwards we have:

Reviewer: ```suggestion
What's more interesting for this post is the `payload` itself. Working backwards we have:
```

> * `ts_ms` is the timestamp of when change happened.,

Reviewer: ```suggestion
* `ts_ms` is the timestamp of when the change happened
```

Reviewer: Better to align on one style (with/without comma/dot), looking at the next bullets

date: 2019-11-01
author: tom_bentley
---

For too long our Kafka Connect story hasn't been quite as "Kubernetes-native" as it could have been.
We had a `KafkaConnect` resource to configure a Kafka Connect _cluster_ but you still had to use the Kafka Connect REST API to actually create a _connector_ within it.
While this wasn't especially difficult using something like `curl`, it stood out because everything else could be done using `kubectl` and it meant that connectors didn't fit into our Kubernetes-native vision.
With the help of a contribution from the community, Strimzi now supports a `KafkaConnector` custom resource and the rest of this blog post is going to explain how to use it using [Debezium](https://debezium.io) as an example. 
As if that wasn't enough, there's some awesome ASCII art to help explain how it all fits together. So read on to find out about this new Kubernetes-native way of managing connectors.

<!--more-->

# Debezi-what?

If you've not heard of Debezium before it is an open source project for applying the [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture) (CDC) pattern to your applications using Kafka.

> But what is CDC? You probable have several databases in your organization; silos full of business-critical data.

Reviewer 3: ```suggestion
But what is CDC? You probably have several databases in your organization; silos full of business-critical data.
```

While you can query those databases any time you want, the essence of your business revolves around how that data gets modified.

> Those modifications trigger, and are triggered by, real world business events (e.g. a phone call from a customer, or a successful payment or whatever)

Reviewer 3: ```suggestion
Those modifications trigger, and are triggered by, real world business events (e.g. a phone call from a customer, a successful payment, or whatever).
```

In order to reflect this, modern application architectures are frequently event-based. 
And so asking for the _current state_ of some tables is not enough; what these architectures need is a stream of _events representing modifications_ to those tables.
That's what CDC is: Capturing the changes to the state data as event data.

Concretely, Debezium works with a number of common DBMSs (MySQL, MongoDB, PostgreSQL, Oracle, SQL Server and Cassandra) and runs as a source connector within a Kafka Connect cluster.

> How Debezium works on the database side depends which database it's using.

Reviewer 3: ```suggestion
How Debezium works on the database side depends on which database it's using.
```

  | sed 's/namespace: .*/namespace: kafka/' \
  | kubectl apply -f - -n kafka 
```

And spin up a Kafka cluster, waiting until it's ready:

```shell
kubectl -n kafka \
    apply -f https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/0.15.0/examples/kafka/kafka-persistent-single.yaml \
  && kubectl wait kafka/my-cluster --for=condition=Ready --timeout=300s -n kafka
```

What we've got so far looks like this

<pre style="line-height: 1.2 !important;"><code>
┌────────────────────────────────┐         ┌────────────────────────────────┐
│ minikube, namespace: kafka     │         │ docker                         │
│                                │         │                                │
│ Kafka                          │         │ MySQL                          │
│  name: my-cluster              │         │                                │
│                                │         └────────────────────────────────┘
│ KafkaConnect                   │
│  name: my-connect              │
│                                │
└────────────────────────────────┘
</code></pre>

> Don't worry, the ASCII art gets a lot better than this!

> What's missing from this picture is Kafka Connect in the minikube box.

Reviewer 3: KafkaConnector in the minikube box.

# Kafka Connect image

The next step is to [create a Strimzi Kafka Connect image](https://strimzi.io/docs/master/#creating-new-image-from-base-str) which includes the Debezium MySQL connector and its dependencies.

First download and extract the Debezium MySQL connector archive
```shell
curl https://repo1.maven.org/maven2/io/debezium/debezium-connector-mysql/0.10.0.Final/debezium-connector-mysql-0.10.0.Final-plugin.tar.gz \
| tar xvz
```

Prepare a `Dockerfile` which adds those connector files to the Strimzi Kafka Connect image

```shell
cat <<EOF >Dockerfile
FROM strimzi/kafka:latest-kafka-2.3.0
USER root:root
EXEC mkdir -p /opt/kafka/plugins/debezium
COPY ./debezium-connector-mysql/ /opt/kafka/plugins/debezium/
USER 1001
EOF
```

Then build the image from that `Dockerfile` and push it to dockerhub.

```shell
# You can use your own dockerhub organization
export DOCKER_ORG=tjbentley
docker build . -t ${DOCKER_ORG}/connect-debezium
docker push ${DOCKER_ORG}/connect-debezium
```

# Create the Connect cluster

Now we can create a `KafkaConnect` cluster in Kubernetes:

```shell

> cat <<EOF | kubectl apply -f -

Reviewer 4: I think this command or the CR needs the namespace setting

apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaConnect
metadata:
  name: my-connect-cluster
  annotations:
  # use-connector-resources configures this KafkaConnect
  # to use KafkaConnector resources to avoid
  # needing to call the Connect REST API directly
    strimzi.io/use-connector-resources: "true"
spec:
  image: ${DOCKER_ORG}/connect-debezium
  replicas: 1
  bootstrapServers: my-cluster-kafka-bootstrap:9093
  tls:
    trustedCertificates:
      - secretName: my-cluster-cluster-ca-cert
        certificate: ca.crt
EOF
```

It's worth pointing out a couple of things about the above resource:

* The `strimzi.io/use-connector-resources: "true"` annotation tells the cluster operator that `KafkaConnector` resources will be used to configure connectors within this Kafka Connect cluster.
* The `spec.image` is the image we just created with `docker`.

# Create the connector

The last piece is to create the `KafkaConnector` resource configured to connect to our "inventory" database in MySQL.

Here's what the `KafkaConnector` resource looks like:

```shell

> cat | kubectl apply -f - << EOF

Reviewer 4: namespace again?

* I'm using `database.hostname: 192.168.99.1` as IP address for connecting to MySQL because I'm using `minikube` with the virtualbox VM driver
  If you're using a different VM driver with `minikube` you might need a different IP address.
* The `database.port: "3306"` works because of the `-p 3306:3306` argument we used when we started up the MySQL server.
* The `database.whitelist: "inventory"` basically tells Debezium to only watch the `inventory` database.
* The `database.history.kafka.topic: "schema-changes.inventory"` configured Debezium to use the `schema-changes.inventory` topic to store the database schema history.

A while after you've created this connector you can have a look at its `status`, using `kubectl get kctr inventory-connector -o yaml`:

```
...
status:
  conditions:
  - lastTransitionTime: "2019-11-18T15:01:53.135Z"
    status: "True"
    type: Ready
  observedGeneration: 1
```

This tells us that the connector is running within the `KafkaConnect` cluster we created in the last step.

> To summarise, we've now the complete picture with the connector talking to MySQL:

Reviewer 3: ```suggestion
To summarise, we now have the complete picture with the connector talking to MySQL:
```

      "snapshot": "false",
      "db": "inventory",
      "table": "customers",
      "server_id": 223344,
      "gtid": null,
      "file": "mysql-bin.000003",
      "pos": 4311,
      "row": 0,
      "thread": 3,
      "query": null
    },
    "op": "u",
    "ts_ms": 1574090237089
  }
}
```

The `schema` object is describing the schema of the actual event payload.

What's more interesting for this post is the `payload` itself. Working backwards we have:
* `ts_ms` is the timestamp of when the change happened
* `op` tells us this was an `u`pdate (an insert would be `c` and a delete would be `d`) 
* the `source` which tells us exactly which table of which database in which server got changed.
* `before` and `after` are pretty self-explanatory, describing the row before and after the update.

You can of course experiment with inserting and deleting rows in different tables (remember, there's a topic for each table).

But what about that `dbserver1` topic which I glossed over? If we look at the messages in there 
(using `k exec my-cluster-kafka-0 -c kafka -i -t -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --topic dbserver1 --from-beginning`) 

> we can see records representing the DDL which was used (when the docker image was created) to create the database.

Reviewer 3: What's the DDL?

As well as the convenience of sticking in the Kubernetes development paradigm, 
the benefits of declarative configuration also apply: 
Your infrastructure is now just code so it can be stored in version control, be peer reviewed and be deployed along with the application that needs it; 
all of this while keeping your organisation's existing tools and processes.

Which is great, but for too long our Kafka Connect story wasn't quite as "Kubernetes-native" as it could have been.
While we had a `KafkaConnect` resource to configure a Kafka Connect _cluster_ you still had to use the Kafka Connect REST API to actually create a _connector_ within it.
While this wasn't especially difficult using something like `curl`, it stood out because everything else could be done using `kubectl` and it meant that connectors didn't fit into our Kubernetes-native vision.

With the help of a contribution from the community, Strimzi now supports a `KafkaConnector` custom resource and the rest of this blog post is going to explain how to use it using [Debezium](https://debezium.io) as an example. 
And as if that wasn't enough, there's some awesome ASCII art.

> <!--more-->

Reviewer 5: The perex is way too long. It looks weird on the weesite. I think you should make it just one paragraph.
